File: archer/core/variables.py
"""Variable processing utilities for single and multipart templates."""

import re
import logging
from typing import Dict, Optional, Tuple, Any

logger = logging.getLogger(__name__)

# Compile the pattern once at module level
VARIABLE_PATTERN = re.compile(r'\$\{([^}]+)\}')


def inject_variables(content: str, variables: Dict[str, str]) -> str:
    """Inject variables into content string."""
    if not content:
        return content
    
    def replace_variable(match):
        var_name = match.group(1)
        if var_name in variables:
            return variables[var_name]
        else:
            logger.warning(f"Variable '{var_name}' not found in provided variables")
            return match.group(0)  # Return original placeholder if variable not found
    
    return VARIABLE_PATTERN.sub(replace_variable, content)


def mask_variables(content: str) -> str:
    """Mask all variables in content for safe logging."""
    if not content:
        return content
    
    def mask_variable(match):
        var_name = match.group(1)
        return f"${{***{var_name}_MASKED***}}"
    
    return VARIABLE_PATTERN.sub(mask_variable, content)


def process_headers(headers: Dict[str, str], variables: Dict[str, str]) -> Tuple[Dict[str, str], Dict[str, str]]:
    """Process headers for both request use and masked logging."""
    request_headers = {}
    masked_headers = {}
    
    for key, value in headers.items():
        request_headers[key] = inject_variables(value, variables)
        masked_headers[key] = mask_variables(value)
    
    return request_headers, masked_headers


def process_query_params(query_params: Optional[Dict[str, str]], variables: Dict[str, str]) -> Tuple[Optional[Dict[str, str]], Optional[Dict[str, str]]]:
    """Process query parameters for both request use and masked logging."""
    if not query_params:
        return None, None
    
    request_params = {}
    masked_params = {}
    
    for key, value in query_params.items():
        request_params[key] = inject_variables(value, variables)
        masked_params[key] = mask_variables(value)
    
    return request_params, masked_params


def process_url(url: str, variables: Dict[str, str]) -> Tuple[str, str]:
    """Process URL for both request use and masked logging."""
    request_url = inject_variables(url, variables)
    masked_url = mask_variables(url)
    return request_url, masked_url


def process_data(data: Optional[str], variables: Dict[str, str]) -> Tuple[Optional[str], Optional[str]]:
    """Process data string for both request use and masked logging."""
    if not data:
        return None, None
    
    request_data = inject_variables(data, variables)
    masked_data = mask_variables(data)
    return request_data, masked_data


def process_json_data(json_data: Optional[Dict[str, Any]], variables: Dict[str, str]) -> Tuple[Optional[Dict[str, Any]], Optional[Dict[str, Any]]]:
    """Process JSON data for both request use and masked logging."""
    if not json_data:
        return None, None
    
    def process_json_recursive(obj: Any) -> Tuple[Any, Any]:
        if isinstance(obj, str):
            request_value = inject_variables(obj, variables)
            masked_value = mask_variables(obj)
            return request_value, masked_value
        elif isinstance(obj, dict):
            request_dict = {}
            masked_dict = {}
            for key, value in obj.items():
                req_val, mask_val = process_json_recursive(value)
                request_dict[key] = req_val
                masked_dict[key] = mask_val
            return request_dict, masked_dict
        elif isinstance(obj, list):
            request_list = []
            masked_list = []
            for item in obj:
                req_item, mask_item = process_json_recursive(item)
                request_list.append(req_item)
                masked_list.append(mask_item)
            return request_list, masked_list
        else:
            # For non-string values (numbers, bools, etc.), return as-is
            return obj, obj
    
    return process_json_recursive(json_data)


def get_variables_from_template_content(content: str) -> set:
    """Extract all variable names from template content."""
    if not content:
        return set()
    
    matches = VARIABLE_PATTERN.findall(content)
    return set(matches)


def validate_variables_provided(required_variables: list, provided_variables: Dict[str, str]) -> list:
    """Validate that all required variables are provided. Returns list of missing variables."""
    if not required_variables:
        return []
    
    missing = []
    for var in required_variables:
        if var not in provided_variables or not provided_variables[var].strip():
            missing.append(var)
    
    return missing


def parse_var_args(var_args: list) -> Dict[str, str]:
    """Parse --var key=value arguments into a dictionary."""
    variables = {}
    
    for var_arg in var_args:
        if '=' not in var_arg:
            raise ValueError(f"Invalid variable format: '{var_arg}'. Use --var key=value")
        
        # Split only on first '=' to allow '=' in values
        key, value = var_arg.split('=', 1)
        
        # Validate key format (kebab-case)
        if not re.match(r'^[a-z][a-z0-9-]*$', key):
            raise ValueError(f"Variable name '{key}' must be in kebab-case format (e.g., 'api-token', 'base-url')")
        
        # Convert kebab-case to UPPER_SNAKE_CASE
        upper_snake_key = key.upper().replace('-', '_')
        variables[upper_snake_key] = value
    
    return variables


def format_var_name_for_cli(upper_snake_name: str) -> str:
    """Convert UPPER_SNAKE_CASE to kebab-case for CLI display."""
    return upper_snake_name.lower().replace('_', '-')

File: internal/variables/variables.go
package variables

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/theinfosecguy/archer/internal/constants"
	"github.com/theinfosecguy/archer/internal/errors"
)

// InjectVariables replaces variable placeholders in content with actual values
func InjectVariables(content string, variables map[string]string) string {
	if content == "" {
		return content
	}

	return constants.VariablePattern.ReplaceAllStringFunc(content, func(match string) string {
		// Extract variable name from ${VAR_NAME}
		varName := match[2 : len(match)-1] // Remove ${ and }
		if value, ok := variables[varName]; ok {
			return value
		}
		// Return original placeholder if variable not found
		return match
	})
}

// MaskVariables masks all variable values in content for safe logging
func MaskVariables(content string) string {
	if content == "" {
		return content
	}

	return constants.VariablePattern.ReplaceAllStringFunc(content, func(match string) string {
		// Extract variable name from ${VAR_NAME}
		varName := match[2 : len(match)-1] // Remove ${ and }
		return fmt.Sprintf("%s%s%s", constants.MaskedVariablePrefix, varName, constants.MaskedVariableSuffix)
	})
}

// ProcessHeaders processes headers for both request use and masked logging
func ProcessHeaders(headers map[string]string, variables map[string]string) (map[string]string, map[string]string) {
	requestHeaders := make(map[string]string)
	maskedHeaders := make(map[string]string)

	for key, value := range headers {
		requestHeaders[key] = InjectVariables(value, variables)
		maskedHeaders[key] = MaskVariables(value)
	}

	return requestHeaders, maskedHeaders
}

// ProcessQueryParams processes query parameters for both request use and masked logging
func ProcessQueryParams(queryParams map[string]string, variables map[string]string) (map[string]string, map[string]string) {
	if queryParams == nil {
		return nil, nil
	}

	requestParams := make(map[string]string)
	maskedParams := make(map[string]string)

	for key, value := range queryParams {
		requestParams[key] = InjectVariables(value, variables)
		maskedParams[key] = MaskVariables(value)
	}

	return requestParams, maskedParams
}

// ProcessURL processes URL for both request use and masked logging
func ProcessURL(url string, variables map[string]string) (string, string) {
	requestURL := InjectVariables(url, variables)
	maskedURL := MaskVariables(url)
	return requestURL, maskedURL
}

// ProcessData processes data string for both request use and masked logging
func ProcessData(data *string, variables map[string]string) (*string, *string) {
	if data == nil {
		return nil, nil
	}

	requestData := InjectVariables(*data, variables)
	maskedData := MaskVariables(*data)
	return &requestData, &maskedData
}

// ProcessJSONData processes JSON data for both request use and masked logging
func ProcessJSONData(jsonData map[string]any, variables map[string]string) (map[string]any, map[string]any) {
	if jsonData == nil {
		return nil, nil
	}

	requestData := processJSONRecursive(jsonData, variables, false).(map[string]any)
	maskedData := processJSONRecursive(jsonData, variables, true).(map[string]any)

	return requestData, maskedData
}

func processJSONRecursive(obj any, variables map[string]string, mask bool) any {
	switch v := obj.(type) {
	case string:
		if mask {
			return MaskVariables(v)
		}
		return InjectVariables(v, variables)

	case map[string]any:
		result := make(map[string]any)
		for key, value := range v {
			result[key] = processJSONRecursive(value, variables, mask)
		}
		return result

	case []any:
		result := make([]any, len(v))
		for i, item := range v {
			result[i] = processJSONRecursive(item, variables, mask)
		}
		return result

	default:
		// For non-string values (numbers, bools, etc.), return as-is
		return obj
	}
}

// GetVariablesFromTemplateContent extracts all variable names from template content
func GetVariablesFromTemplateContent(content string) []string {
	if content == "" {
		return nil
	}

	matches := constants.VariablePattern.FindAllStringSubmatch(content, -1)
	varSet := make(map[string]bool)

	for _, match := range matches {
		if len(match) > 1 {
			varSet[match[1]] = true
		}
	}

	vars := make([]string, 0, len(varSet))
	for v := range varSet {
		vars = append(vars, v)
	}

	return vars
}

// ValidateVariablesProvided validates that all required variables are provided
// Returns list of missing variables
func ValidateVariablesProvided(requiredVariables []string, providedVariables map[string]string) []string {
	if len(requiredVariables) == 0 {
		return nil
	}

	missing := make([]string, 0)
	for _, varName := range requiredVariables {
		value, ok := providedVariables[varName]
		if !ok || strings.TrimSpace(value) == "" {
			missing = append(missing, varName)
		}
	}

	return missing
}

// ParseVarArgs parses --var key=value arguments into a dictionary
func ParseVarArgs(varArgs []string) (map[string]string, error) {
	variables := make(map[string]string)

	for _, varArg := range varArgs {
		if !strings.Contains(varArg, constants.VariableSeparator) {
			return nil, &errors.ValidationError{
				Message: fmt.Sprintf(constants.InvalidVariableFormat, varArg),
			}
		}

		// Split only on first '=' to allow '=' in values
		parts := strings.SplitN(varArg, constants.VariableSeparator, 2)
		key := parts[0]
		value := parts[1]

		// Validate key format (kebab-case)
		if !constants.KebabCasePattern.MatchString(key) {
			return nil, &errors.ValidationError{
				Message: fmt.Sprintf(constants.InvalidKebabCase, key),
			}
		}

		// Convert kebab-case to UPPER_SNAKE_CASE
		upperSnakeKey := strings.ToUpper(strings.ReplaceAll(key, constants.KebabToSnakeSeparator, constants.SnakeCaseSeparator))
		variables[upperSnakeKey] = value
	}

	return variables, nil
}

// FormatVarNameForCLI converts UPPER_SNAKE_CASE to kebab-case for CLI display
func FormatVarNameForCLI(upperSnakeName string) string {
	return strings.ToLower(strings.ReplaceAll(upperSnakeName, constants.SnakeCaseSeparator, constants.KebabToSnakeSeparator))
}

// MarshalJSONWithVariables marshals data to JSON string with variable injection
func MarshalJSONWithVariables(data map[string]any, variables map[string]string) (string, error) {
	processed := processJSONRecursive(data, variables, false)
	bytes, err := json.Marshal(processed)
	if err != nil {
		return "", err
	}
	return string(bytes), nil
}
